### PR TITLE
Release announcement for Kroxylicious 0.4.0

### PR DESCRIPTION
**Pull request by k-wall**

### CODE REVIEW

File: _posts/2023-12-13-release-0_4_0.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+categories:  releases kroxylcious topic-encryption
+---
+
+The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.   The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers.
+
+The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
+
+The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  Other Key Management System integrations are planned soon.
```
Comment (SamBarker) [inline _posts/2023-12-13-release-0_4_0.md:16]:
```suggestion
The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  We will provide other selected Key Management System integrations soon, however we are also leveraging our plug-in architecture to allow users to supply their own integrations.
```

File: _posts/2023-12-13-release-0_4_0.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
+---
+layout: post
+title:  "kroxylicious release 0.4.0"
+date:   2023-12-14 00:00:00 +1300
+author: "Keith Wall"
+author_url: "https://www.github.com/k-wall"
+categories:  releases kroxylcious topic-encryption
+---
+
+The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.   The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers.
```
Comment (SamBarker) [inline _posts/2023-12-13-release-0_4_0.md:10]: I'm in two minds about this, its true, but I worry it makes it sound like its finished. Saying its the first iteration isn't very helpful either as we don't actually expect the the filter to change most of the remaining work is around the non functional requirements or reliability. 

Maybe adding another sentence along the lines of 
> We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0](www.github.com/kroxylicious/...) milestone for details

File: _posts/2023-12-13-release-0_4_0.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
+
+The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  Other Key Management System integrations are planned soon.
+
+The diagram below shows how the system works at a high level.
+
+![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/02b6c1fd-c8a4-495a-80a1-0c7e8f2f97f2)
```
Comment (SamBarker) [inline _posts/2023-12-13-release-0_4_0.md:20]: awesome diagram!

File: _posts/2023-12-13-release-0_4_0.md (hunk 2 above)
Comment (SamBarker) [inline _posts/2023-12-13-release-0_4_0.md:10]: From a functional standpoint this is the big news, from a user standpoint I'd say the binary distribution is probably as big news.

**Review by SamBarker — COMMENTED** [review]
Looks good so far @k-wall I really like the diagram. 

I think we want to mention the binary distribution as well.

File: _posts/2023-12-13-release-0_4_0.md (hunk 2 above)
Comment (k-wall) [reply _posts/2023-12-13-release-0_4_0.md:10]: I agree it is important that we get over this isn't finished and folks shouldn't be dropping into production.  I think your words convey that well.  I'll add them.

File: _posts/2023-12-13-release-0_4_0.md [hunk 4]
```diff
@@ -7,7 +7,9 @@ author_url: "https://www.github.com/k-wall"
 categories:  releases kroxylcious topic-encryption
 ---
 
-The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.   The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers.
+The Kroxylicious project is very pleased to announce the release of Kroxylicious 0.4.0.
+
+The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```
Comment (k-wall) [inline _posts/2023-12-13-release-0_4_0.md:12]: I used EnvelopeEncryption here as that's what we've called the filter - so the argument being consistency.  You could argue that the fact it is using envelope encryption is an implementation detail.  The important thing is what it does - TopicEncryption, or EncryptionAtRest.

Comment (SamBarker) [reply _posts/2023-12-13-release-0_4_0.md:12]: I think you've expressed it well. We are delivering encryption-at-rest via an envelope encryption filter.

File: _posts/2023-12-13-release-0_4_0.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+categories:  releases kroxylcious topic-encryption
+---
+
+The Kroxylicious project is very pleased to announce the [release](https://github.com/kroxylicious/kroxylicious/releases/tag/v0.4.0) of Kroxylicious 0.4.0.
+
+The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
+
+The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
+
+The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.
```
Comment (SamBarker) [inline _posts/2023-12-13-release-0_4_0.md:16]:
```suggestion
The Kafka Cluster *never* sees the plain text of your records. Filter configuration is used to specify which topic(s) should be encrypted by which key.
```

credit to @gracegrimwood for figuring out the right form of words.

File: _posts/2023-12-13-release-0_4_0.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  We will provide other selected Key Management System integrations soon, however we are also leveraging our plug-in architecture to allow users to supply their own integrations.
```
Comment (gracegrimwood) [inline _posts/2023-12-13-release-0_4_0.md:18]: Nit: rephrasing this so it reads a bit better
```suggestion
The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with a [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm) integration.  We will provide other selected Key Management System integrations soon, however users can leverage the plug-in architecture of Kroxylicious to supply their own integrations.
```

File: _posts/2023-12-13-release-0_4_0.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+title:  "kroxylicious release 0.4.0"
+date:   2023-12-14 00:00:00 +1300
+author: "Keith Wall"
+author_url: "https://www.github.com/k-wall"
+categories:  releases kroxylcious topic-encryption
+---
+
+The Kroxylicious project is very pleased to announce the [release](https://github.com/kroxylicious/kroxylicious/releases/tag/v0.4.0) of Kroxylicious 0.4.0.
+
+The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```
Comment (robobario) [inline _posts/2023-12-13-release-0_4_0.md:12]:
```suggestion
The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implementation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```

File: _posts/2023-12-13-release-0_4_0.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+The Kafka Cluster *never* sees the plain text of your records.   You use filter configuration to specify which topic(s) you wish to be encrypted by which key.
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with an integration with [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm).  We will provide other selected Key Management System integrations soon, however we are also leveraging our plug-in architecture to allow users to supply their own integrations.
+
+The diagram below shows how the system works at a high level.
+
+![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/02b6c1fd-c8a4-495a-80a1-0c7e8f2f97f2)
+
+If the use-case interest you, there are some resources (TODO: link) to help you try out the feature in development.  There will be more information landing on the website soon, but in the meanwhile if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).
```
Comment (robobario) [inline _posts/2023-12-13-release-0_4_0.md:24]:
```suggestion
If the use-case interests you, there are some resources (TODO: link) to help you try out the feature in development.  There will be more information landing on the website soon, but in the meanwhile if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).
```

File: _posts/2023-12-13-release-0_4_0.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+author: "Keith Wall"
+author_url: "https://www.github.com/k-wall"
+categories:  releases kroxylcious topic-encryption
+---
+
+The Kroxylicious project is very pleased to announce the [release](https://github.com/kroxylicious/kroxylicious/releases/tag/v0.4.0) of Kroxylicious 0.4.0.
+
+The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for the Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Checkout our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
+
+The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
```
Comment (gracegrimwood) [inline _posts/2023-12-13-release-0_4_0.md:14]: Nit: capitalise Kafka as it is a noun
```suggestion
The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and applies encryption to the Kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
```

File: _posts/2023-12-13-release-0_4_0.md (hunk 5 above)
Comment (gracegrimwood) [reply _posts/2023-12-13-release-0_4_0.md:16]: I have a few more things I would alter here, like using a single space between sentences
```suggestion
The Kafka Cluster *never* sees the plain text of encrypted records. Filter configuration is used to specify which topic(s) are to be encrypted by which key.
```

File: _posts/2023-12-13-release-0_4_0.md (hunk 7 above)
Comment (gracegrimwood) [inline _posts/2023-12-13-release-0_4_0.md:12]: Nits: no "the" before Apache Kafka, and "check out" is always two words unless you're referring to the `git` command
```suggestion
The big news is this release includes our `EnvelopeEncryption` filter which provides an **Encryption-at-Rest** solution for Apache Kafka(tm) which is _transparent_ to both clients and brokers. We have more work planned to iterate on the implentation to reach production quality. Check out our [0.5.0 milestone](https://github.com/kroxylicious/kroxylicious/milestones/0.5.0) for details.
```

**Review by gracegrimwood — COMMENTED** [review]
I have a few nits, but otherwise this looks good. Thanks Keith!

File: _posts/2023-12-13-release-0_4_0.md (hunk 9 above)
Comment (robobario) [inline _posts/2023-12-13-release-0_4_0.md:14]:
```suggestion
The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and encrypts the kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
```

**Review by robobario — APPROVED** [review]
LGTM with nits

**Review by SamBarker — APPROVED** [review]

**Comment by k-wall** [comment]
Maybe the diagram should have some yellow text explaining Vault can be on-premise or cloud.  At the moment it looks like it has to be the customer's network.

**Comment by k-wall** [comment]
So, I'm thinking of linking the blog post to something like this:

https://github.com/kroxylicious/kroxylicious/blob/13997dfa9a20702af3ab7cfd8663b1a76c5e14c5/kubernetes-examples/envelope-encryption/README.md

and:

https://asciinema.org/a/627076

File: _posts/2023-12-13-release-0_4_0.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with a [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm) integration.  We will provide other selected Key Management System integrations soon, however users can leverage the plug-in architecture of Kroxylicious to supply their own integrations.
+
+The diagram below shows how the system works at a high level.
+
+![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/09aeb1f4-f420-4d89-b9f7-394473dddc05)
+
+If this use-case interests you, a [short demo](https://asciinema.org/a/627519) that shows the feature in action will help you understand more and there's some [interim instructions[(https://github.com/kroxylicious/kroxylicious/blob/f5f62a48d1db67309eee1696bb7a70dbb83ba622/kubernetes-examples/envelope-encryption/README.md) to help you try out the feature in development.  There will be more information landing on the website soon, but meanwhile, if you have questions or comments please head over to the [Slack channel](https://kroxylicious.slack.com/).
```
Comment (k-wall) [inline _posts/2023-12-13-release-0_4_0.md:24]: My intent is that https://github.com/kroxylicious/kroxylicious/pull/825 is merged then this link it updated to refer to main.    For the next release https://github.com/kroxylicious/kroxylicious/issues/827 should be resolved, so we'd be able to reference the example on tag (hopefully it'll be superseded by quickstart content by then too).

File: _posts/2023-12-13-release-0_4_0.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+The `EnvelopeEncryption` filter works by intercepting all produce requests from applications and encrypts the Kafka records as they pass through Kroxylicious. On the consume path, the reverse happens - the filter intercepts the fetch responses and decrypts the records before they are sent to the application.
+
+The Kafka Cluster *never* sees the plain text of your records. Filter configuration is used to specify which topic(s) should be encrypted by which key.
+
+The encryption keys themselves are stored safely in a Key Management System.  0.4.0 ships with a [HashiCorp Vault](https://www.hashicorp.com/products/vault)(tm) integration.  We will provide other selected Key Management System integrations soon, however users can leverage the plug-in architecture of Kroxylicious to supply their own integrations.
+
+The diagram below shows how the system works at a high level.
+
+![image](https://github.com/kroxylicious/kroxylicious.github.io/assets/18440250/09aeb1f4-f420-4d89-b9f7-394473dddc05)
```
Comment (k-wall) [inline _posts/2023-12-13-release-0_4_0.md:22]: I've updated the diagram to point out Vault can be either on-premise or cloud.

**Review by SamBarker — APPROVED** [review]
LGTM

**Review by gracegrimwood — APPROVED** [review]
LGTM :)

**Pull request merged**